### PR TITLE
Add Stringer and Stringers method for *Event  context

### PR DESCRIPTION
**Pull request by dennypenta**
Hi.
I made an implementation for adding values as a fmt.Stringer interface. It allows don't create new strings from struct in many cases and do it only if your log level and other conditions require it.

### CODE REVIEW

**Comment by rs** [comment]
I think the `Interface` would have the same effect.

File: event.go [hunk 1]
```diff
@@ -235,6 +235,26 @@ func (e *Event) Strs(key string, vals []string) *Event {
 	return e
 }
 
+// Stringer adds the field key with val as a string to the *Event context.
+// Use Stringer if possible, because it creates string value in lazy mode.
+func (e *Event) Stringer(key string, val fmt.Stringer) *Event {
+	if e == nil {
+		return e
+	}
+	e.buf = enc.AppendString(enc.AppendKey(e.buf, key), val.String())
```
Comment (hummerd) [inline event.go]: This will cause panic if val is nil

**Comment by hummerd** [comment]
> I think the `Interface` method would have the same effect.
```
log.Interface("url", url) // will produce {"url":{"Scheme":"http","Opaque":"","User":null,"Host":"example.com","Path":"/some/path","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":""}}

// While 
log.Stringer("url", url) // will produce {"url": "http://example.com/some/path"}
```

File: event.go (hunk 1 above)
Comment (dennypenta) [reply event.go]: thanks, my miss

Comment (dennypenta) [reply event.go]: pushed the fix

**Comment by dennypenta** [comment]
If you ok with this feature, I'm going to write tests

**Comment by dennypenta** [comment]
I added tests, but it seems like tests for tests, probably it will be better to remove them.

File: event.go [hunk 2]
```diff
@@ -235,6 +235,30 @@ func (e *Event) Strs(key string, vals []string) *Event {
...
 }
 
+// Stringer adds the field key with val as a string to the *Event context.
+// Use Stringer if possible, because it creates string value in lazy mode.
+func (e *Event) Stringer(key string, val fmt.Stringer) *Event {
+	if e == nil {
+		return e
+	}
+	if val == nil {
+		e.buf = enc.AppendString(enc.AppendKey(e.buf, key), "nil")
```
Comment (rs) [inline event.go]: Should be “null”.

File: internal/json/string_test.go [hunk 3]
```diff
@@ -72,6 +100,15 @@ func TestAppendString(t *testing.T) {
...
 }
 
+func TestAppendStringer(t *testing.T) {
+	for _, tt := range encodeStringerTests {
+		b := enc.AppendString([]byte{}, tt.in.String())
+		if got, want := string(b), tt.out; got != want {
+			t.Errorf("appendString(%q) = %#q, want %#q", tt.in.String(), got, want)
+		}
+	}
+}
```
Comment (rs) [inline internal/json/string_test.go]: Please add a test for nil case.

File: event.go (hunk 2 above)
Comment (hummerd) [inline event.go]: You can use e.buf = enc.AppendInterface(enc.AppendKey(e.buf, key), nil) to achieve that

File: internal/json/string.go [hunk 4]
```diff
@@ -29,6 +32,33 @@ func (e Encoder) AppendStrings(dst []byte, vals []string) []byte {
...
+	if vals[0] == nil {
+		dst = e.AppendInterface(dst, nil)
+	} else {
+		dst = e.AppendString(dst, vals[0].String())
+	}
+
+	if len(vals) > 1 {
+		for _, val := range vals[1:] {
+			if val == nil {
+				dst = e.AppendString(append(dst, ','), "nil")
```
Comment (rs) [inline internal/json/string.go]: `nil` is not valid in json.

**Comment by dennypenta** [comment]
Guys, I made an update month ago.
If I need a little bit fixes - let me know pls.

File: event.go [hunk 5]
```diff
@@ -235,6 +235,30 @@ func (e *Event) Strs(key string, vals []string) *Event {
...
+// Use Stringer if possible, because it creates string value in lazy mode.
+func (e *Event) Stringer(key string, val fmt.Stringer) *Event {
+	if e == nil {
+		return e
+	}
+	if val == nil {
+		e.buf = enc.AppendInterface(enc.AppendKey(e.buf, key), nil)
+		return e
+	}
+	e.buf = enc.AppendString(enc.AppendKey(e.buf, key), val.String())
```
Comment (rs) [inline event.go]: Either this implementation should go to an `AppendStringer` or `AppendStringers` should be moved in this file. It is odd to have this feature implemented in two places.

Comment (dennypenta) [reply event.go]: I'm not sure what are you talking about. 
You propose to implement `AppendStriner` for encoder, but it will whole copy of `AppendString` . Or I can call `AppendString ` inside the `AppendStriner `

**Comment by dennypenta** [comment]
Good day.
Guys, what is the plan?

File: event.go (hunk 5 above)
Comment (rs) [reply event.go]: My point is that you repeat the `if val == nil then AppendInterface else AppendString` in many places where it should be a `AppendStringer` in the encoders instead.
